### PR TITLE
[kafka] Round-2: LR warmup + cosine decay on 6L base

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The current 6L baseline uses constant `lr=2e-4` throughout training. With only
3-4 epochs in budget, the learning rate schedule matters enormously — each epoch
represents ~25% of total training. A linear warmup followed by cosine decay is
standard practice for transformer training (ViT, Swin, DeiT) and has two benefits:

1. **Warmup**: prevents large gradient updates in early steps when model weights
   are far from their final values, reducing the risk of divergence on the first epoch.
2. **Cosine decay**: allows the model to refine predictions with smaller updates in
   later epochs, reducing noise in the final EMA checkpoint.

The human researcher team (Issue #18) specifically called out convergence speed as
a priority — this addresses it directly by giving each epoch its best possible LR
rather than a fixed rate that may be too high late and too low early.

**This requires adding an LR scheduler to `train.py` — a small code change.**

## Instructions

**1. Add config fields to `Config`:**

```python
lr_warmup_steps: int = 0       # steps for linear warmup (0 = no warmup)
lr_cosine_decay: bool = False   # cosine decay from lr to lr_min after warmup
lr_min: float = 1e-6            # minimum LR at end of cosine decay
```

**2. Add a scheduler helper** (after the optimizer is constructed):

```python
def get_lr_scale(step: int, warmup_steps: int, total_steps: int,
                 cosine_decay: bool, lr_min: float, lr_max: float) -> float:
    """Compute LR multiplicative scale at a given step."""
    if step < warmup_steps:
        return step / max(warmup_steps, 1)  # linear warmup
    if not cosine_decay:
        return 1.0  # constant after warmup
    # Cosine decay from lr_max -> lr_min
    progress = (step - warmup_steps) / max(total_steps - warmup_steps, 1)
    cos_scale = (1 + math.cos(math.pi * progress)) / 2.0
    return lr_min / lr_max + cos_scale * (1 - lr_min / lr_max)
```

`math` is already imported. Use `total_steps = 50000` as a conservative estimate
(covers 3-4 epochs at 6L throughput of ~2.1 it/s x 270 min training = ~34k steps).

**3. Apply the scheduler in the training loop**, after each optimizer step:

```python
if config.lr_warmup_steps > 0 or config.lr_cosine_decay:
    scale = get_lr_scale(
        global_step, config.lr_warmup_steps,
        total_steps=50000,
        cosine_decay=config.lr_cosine_decay,
        lr_min=config.lr_min, lr_max=config.lr,
    )
    for pg in optimizer.param_groups:
        pg["lr"] = config.lr * scale
    if global_step % config.gradient_log_every == 0:
        wandb.log({"train/lr": optimizer.param_groups[0]["lr"]}, step=global_step)
```

**4. Add CLI flag wiring:**

```python
parser.add_argument("--lr-warmup-steps", type=int, default=0)
parser.add_argument("--lr-cosine-decay", action="store_true", default=False)
parser.add_argument("--no-lr-cosine-decay", dest="lr_cosine_decay", action="store_false")
parser.add_argument("--lr-min", type=float, default=1e-6)
```

**5. Run 3 arms** (3 GPUs):

| GPU | `--lr-warmup-steps` | `--lr-cosine-decay` | `--wandb-name` |
|---|---:|---|---|
| 0 | 1000 | no | `kafka-warmup1k` |
| 1 | 1000 | yes | `kafka-warmup1k-cosine` |
| 2 | 500  | yes | `kafka-warmup500-cosine` |

1000 steps is ~8 min of warmup at 6L throughput; 500 steps is ~4 min.

**Run command (template):**

```bash
cd target/
python train.py \
  --lr-warmup-steps <STEPS> \
  [--lr-cosine-decay] \
  --lr-min 1e-6 \
  --model-layers 6 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --volume-loss-weight 2.0 \
  --batch-size 8 \
  --validation-every 1 \
  --lr 2e-4 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --ema-decay-start 0.99 --ema-decay-end 0.9999 \
  --clip-grad-norm 1.0 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms \
  --wandb-group kafka-lr-schedule-6l \
  --wandb-name kafka-<SLUG>
```

**Key diagnostics to watch:**
- `train/lr` — confirm warmup + decay shape is correct.
- `train/grad/pre_clip_norm` in first 2000 steps — should be lower with warmup.
- Per-epoch val trajectory — does warmup improve epoch-1 quality vs baseline 22.78?

## Baseline (current yi best — PR #14 senku 6L/256d, constant lr=2e-4)

| Metric | 6L yi best (`et4ajeqj`) | AB-UPT |
|---|---:|---:|
| `test_primary/abupt_axis_mean_rel_l2_pct` | **13.15** | — |
| `test_primary/surface_pressure_rel_l2_pct` | **7.64** | 3.82 |
| `test_primary/wall_shear_rel_l2_pct` | **13.47** | 7.29 |
| `test_primary/volume_pressure_rel_l2_pct` | **13.58** | 6.08 |
| `test_primary/wall_shear_x_rel_l2_pct` | **11.53** | 5.35 |
| `test_primary/wall_shear_y_rel_l2_pct` | **16.23** | 3.65 |
| `test_primary/wall_shear_z_rel_l2_pct` | **16.75** | 3.63 |

Per-epoch trajectory (6L baseline): 22.78->15.89->13.30->12.36 (partial epoch 4).

## Results

Add a PR comment with:
1. For each arm: all `test_primary/*_rel_l2_pct` and `full_val_primary/*`.
2. `train/lr` curve screenshots or W&B link.
3. Per-epoch val trajectory vs baseline (22.78 at epoch 1).
4. `train/grad/pre_clip_norm` in first 2000 steps.
5. W&B run IDs and URLs.

## Constraints

- Do **not** modify `data/*`, `pyproject.toml`, or `instructions/*`.
- Do **not** override `SENPAI_MAX_EPOCHS` or `SENPAI_TIMEOUT_MINUTES`.
- `test_primary/*` must **not** be NaN.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #67 because ADVISOR: Closing as stale orphan. This PR (Round-2 LR warmup + cosine decay) was assigned to kafka but never started — no kafka student pod exists on the yi branch, and the PR has had zero progress for 35+ hours. The hypothesis is now superseded by PR #99 (fern Round-3 LR peak sweep on thorfinn base) which directly explores LR scheduling on the current best baseline. No work is lost; closing to keep the board clean..
